### ISSUE TITLE
segfault after several thousand calls to detectFile

### ISSUE THREAD

**Issue by michael-ts** [mscdex/mmmagic#86]
I'm trying to index a drive and have file system walker in which I call detectFile to get the MIME type.  I have tried initializing using both before each call to detectFile, and only once at the beginning:

`var magic = new Magic("./magic.mgc",mmm.MAGIC_MIME_TYPE)`

What I see happen is after the first 135,167 calls to detectFile(), node.js crashes:

`Segmentation fault: 11`

I modified my walker to only start calling detectFile after the last file that succeeded.  However, subsequently fewer number of calls succeed before a segfault occurs:

Second batch: 8869
Third batch: 1283
Fourth batch: 0

The file that the initial attempt fails on succeeds when running detectFile directly on the same file.

However, using the `segfault-handler` module it does appear that it is magic that is crashing.

```
PID 78463 received SIGSEGV for address: 0x104fda239
0   segfault-handler.node               0x0000000103474f1b _ZL16segfault_handleriP9__siginfoPv + 283
1   libsystem_platform.dylib            0x00007fff8d579f1a _sigtramp + 26
2   ???                                 0xb43c9f9dd34de29e 0x0 + 12987430925633446558
3   magic.node                          0x0000000103400045 der_cmp + 81
4   magic.node                          0x0000000103409503 magiccheck + 1019
5   magic.node                          0x0000000103407a58 match + 1024
6   magic.node                          0x000000010340763a file_softmagic + 178
7   magic.node                          0x000000010340193d file_buffer + 728
8   magic.node                          0x0000000103402901 file_or_fd + 562
9   magic.node                          0x00000001033f7bee _ZN5Magic10DetectWorkEP9uv_work_s + 104
10  node                                0x0000000100653e86 worker + 90
11  node                                0x000000010065e895 uv__thread_start + 25
12  libsystem_pthread.dylib             0x00007fff9242d05a _pthread_body + 131
13  libsystem_pthread.dylib             0x00007fff9242cfd7 _pthread_body + 0
14  libsystem_pthread.dylib             0x00007fff9242a3ed thread_start + 13

```

Monitoring the list of files open by node and the process statistics doesn't reveal any runaway use of resources.

**Comment by michael-ts** [comment]
At der.c, line 270, the value of _len_ before the segfault is hugely negative.

**Comment by michael-ts** [comment]
In softmagic.c, line 1185, _offset_ is excessively large.

**Comment by michael-ts** [comment]
Ok, I've tracked the issue down as far as the _struct magic_set_ sub-struct _cont_.  At the point right before the error occurs, the particular instance of magic_set has a _struct cont_ with len=11, and several level_info[].off values are ridiculously large.  However, I can't find any instance where these values are being assigned, which leads me to believe that corruption is coming from some place else.

And yes, I realize that this is in upstream code, but I haven't yet figured out how to interact with the upstream code, the github code is a read-only mirror and what not; it wasn't too hard for me to figure out how to clone this project and play around with trying to debug it.

**Comment by michael-ts** [comment]
If my understanding is correct, I can't explicitly destruct a Magic object.  Even if the variable goes out of scope, I'm not sure if it is being destructed or if there is underlying state that is being reused.

**Comment by michael-ts** [comment]
Maybe something un-initialized somewhere?  I can run multiple invocations of the same code where the detectFile on the first walked file crashes, and at some point it will start working.

**Comment by michael-ts** [comment]
Uninitialized data appears to be confirmed.   I replaced all occurrences of `malloc(%s)` in apprentice.c with `calloc(1,%s)` and I am no longer able to reproduce the issue.

**Comment by returnString** [comment]
Am also seeing this, albeit with far fewer calls to detectFile.

```
#0  0x00007f623c110be8 in der_cmp () from /home/ruan/happy/node_modules/mmmagic/build/Release/magic.node
#1  0x00007f623c108c05 in magiccheck () from /home/ruan/happy/node_modules/mmmagic/build/Release/magic.node
#2  0x00007f623c10c5a4 in match () from /home/ruan/happy/node_modules/mmmagic/build/Release/magic.node
#3  0x00007f623c10c831 in file_softmagic () from /home/ruan/happy/node_modules/mmmagic/build/Release/magic.node
#4  0x00007f623c104b02 in file_buffer () from /home/ruan/happy/node_modules/mmmagic/build/Release/magic.node
#5  0x00007f623c0fc8ba in magic_buffer () from /home/ruan/happy/node_modules/mmmagic/build/Release/magic.node
#6  0x00007f623c0f9f1d in Magic::DetectWork(uv_work_s*) () from /home/ruan/happy/node_modules/mmmagic/build/Release/magic.node
#7  0x0000000000f90c81 in worker (arg=arg@entry=0x0) at ../deps/uv/src/threadpool.c:95
#8  0x0000000000f9f8b9 in uv__thread_start (arg=<optimised out>) at ../deps/uv/src/unix/thread.c:49
#9  0x00007f623fa51184 in start_thread (arg=0x7f622e7fc700) at pthread_create.c:312
#10 0x00007f623f77e37d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

We've got 10 or so files that get sent through `detect` in our proper app. I've got it down to a test case of three, two using Latin characters and the third using Cyrillic; the segfault always occurs on the Cyrillic file specifically. Will try and provide a minimal repro later.

**Comment by returnString** [comment]
I've attached [my minimal repro](https://github.com/mscdex/mmmagic/files/349390/archive.tar.gz).

Triggers the segfault on about 50% of runs for me. Just run `npm install` beforehand. We're using Node 4.4.3.

**Comment by mscdex** [comment]
@returnString Thanks for the code, I was able to reproduce the error with it. The upgrade to libmagic 5.28 in 48316fd1cd62ca5f61c2a272571bdc2c243cf56a seems to fix the segfault.

Thanks @michael-ts and @returnString for digging into this!
